### ISSUE TITLE
Failed to deserialize

### ISSUE THREAD

**Issue by leontoeides** [vincent-herlemont/native_db#139]
Good morning,

Thank you for this crate! It's exactly what I've been looking for. The way you've handled secondary indicies is perfect 👌and very powerful.

Unfortunately when I use the crate I get this error:

```
Failed to deserialize the struct #struct_name
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

This happens when I try to scan the entire database.

```
        rtxn
            .scan()
            .primary()?
            .all()
            .for_each(|my_type: MyType| { /* my code here */ });
```

I might be able to provide more details later, I just wanted to make sure I opened this issue before I forgot

**Comment by vincent-herlemont** [comment]
@leontoeides Thank you for reporting this issue. The error message is unclear, and a first improvement has just been merged https://github.com/vincent-herlemont/native_db/pull/140. And secondly, instead of causing the program to crash, a proper error will be returned.

In my opinion, you are storing data in your database and evolving your models without splitting them into different versions. This can often happen when you are developing an application. If you delete your database and re-run your code, does the problem still occur?

**Comment by leontoeides** [comment]
I'll try to provide more details, eventually, when I can. 

In my case, the struct definition has not changed. I was working on a brand new database, and I was reading back all records I'd just written from a blank database.

The struct I'm using is pretty hefty and I'd imagine it'd be challenging case for a serializer/deserializer. I am successfully using this same struct with `rmp` in a different project, though. The struct includes others structs and collections. ~40 records weighed in at about 1.5 Mb in the `redb` file.

**Comment by vincent-herlemont** [comment]
@leontoeides Ok, so under the hood, `native_db` uses [`native_model`](https://github.com/vincent-herlemont/native_model) to serialize/deserialize these types. By default, `native_model` uses [bincode/1.3.3](https://docs.rs/bincode/1.3.3/bincode/index.html) with `serde`. In order to see where the problem come from, can you please perform the following tests:
- Try to encode/decode your data without `native_db` using `native_model`’s [`encode`](https://docs.rs/native_model/latest/native_model/fn.encode.html) / [`decode`](https://docs.rs/native_model/latest/native_model/fn.decode.html)
- Try to encode/decode your data with [bincode/1.3.3](https://docs.rs/bincode/1.3.3/bincode/index.html) and `serde`.
And see if there is an error.

If the error persists and you find no solution, you can always use `rmp-serde` with `native_model` by using the custom serialization/deserialization methods described here: [`native_model`/Custom serialization format](https://github.com/vincent-herlemont/native_model/blob/60ccc5fdcd00b17b7c4b9e76c72b1a83fe31b5b0/README.md#custom-serialization-format).

**Comment by dylan-bowker** [comment]
Both `native_model` and `bincode` failed to round-trip my struct. 

`bincode`'s error was `DeserializeAnyNotSupported` or `Bincode does not support the serde::Deserializer::deserialize_any method`.

I remember this one. I recall playing with `bincode` before. I had to move away from it because it doesn't work with my data. I'd suggest you do the same. `musli::storage` looks interesting but there's no `serde` support.

`postcard` failed with `DecodeBodyError(DecodeError { msg: "This is a feature that PostCard will never implement", source: This is a feature that PostCard will never implement`

Just some feedback that the struct for postcard was kind of hard to find.
`#[native_model(id = 1, version = 1, with = native_model::postcard_1_0::PostCard)]`

I implemented the an encoder/decoder for `rmp-serde` and it works well for me. Very clever setup. Your crate is full of surprises. It seems like you've thought of everything!

```
pub struct RmpSerde;

impl<T: Serialize> native_model::Encode<T> for RmpSerde {
    type Error = rmp_serde::encode::Error;
    fn encode(obj: &T) -> Result<Vec<u8>, Self::Error> {
        rmp_serde::encode::to_vec(obj)
    }
}

impl<T: for<'a> Deserialize<'a>> native_model::Decode<T> for RmpSerde {
    type Error = rmp_serde::decode::Error;
    fn decode(data: Vec<u8>) -> Result<T, Self::Error> {
        rmp_serde::decode::from_slice(&data)
    }
}
```

**Comment by leontoeides** [comment]
Both `native_model` and `bincode` failed to round-trip my struct. 

`bincode`'s error was `DeserializeAnyNotSupported` or `Bincode does not support the serde::Deserializer::deserialize_any method`.

I remember this one. I recall playing with `bincode` before. I had to move away from it because it doesn't work with my data. You may want to do the same (bincode as a default) for this same reason.

[musli::storage](https://docs.rs/musli/latest/musli/storage/) looks interesting but there's no `serde` support.

Also, `postcard` failed with `DecodeBodyError(DecodeError { msg: "This is a feature that PostCard will never implement", source: This is a feature that PostCard will never implement`

Just some feedback - the struct for postcard was hard to find because it didn't make it into the [docs](https://docs.rs/native_model). It seems like there's a way to [document featured-gated items](https://github.com/rust-lang/cargo/issues/8103), though.

This was tricky to find:
`#[native_model(id = 1, version = 1, with = native_model::postcard_1_0::PostCard)]`

I implemented a codec using `rmp-serde` (see below) and it works well for me. The encoder/decoder trait + macro setup is very clever! Your crate is full of surprises. It seems like you've thought of everything.

```
pub struct RmpSerde;

impl<T: Serialize> native_model::Encode<T> for RmpSerde {
    type Error = rmp_serde::encode::Error;
    fn encode(obj: &T) -> Result<Vec<u8>, Self::Error> {
        rmp_serde::encode::to_vec(obj)
    }
}

impl<T: for<'a> Deserialize<'a>> native_model::Decode<T> for RmpSerde {
    type Error = rmp_serde::decode::Error;
    fn decode(data: Vec<u8>) -> Result<T, Self::Error> {
        rmp_serde::decode::from_slice(&data)
    }
}
```

**Comment by vincent-herlemont** [comment]
> Just some feedback - the struct for postcard was hard to find because it didn't make it into the [docs](https://docs.rs/native_model). It seems like there's a way to https://github.com/rust-lang/cargo/issues/8103, though.

Thanks to pointing it, I have opened an issue here https://github.com/vincent-herlemont/native_model/issues/70.

---

@leontoeides It might be a good idea to add `rmp-serde` as another available decoder with a feature. What do you think? Feel free to make a PR if you wish.

---

@leontoeides Is everything good for you on this topic? If so, I will let you close the issue.

**Comment by LunaeSomnia** [comment]
Hi. I'm also experiencing this kind of issues in my project. @leontoeides, could you specify what exactly you did to fix the issue on your end? Thanks.

**Comment by leontoeides** [comment]
> Hi. I'm also experiencing this kind of issues in my project. @leontoeides, could you specify what exactly you did to fix the issue on your end? Thanks.

Hello @LunaeSomnia, it looks like the default serializer `bincode` isn't feature complete so it's crashing. Try using `rmp-serde` instead, which is a [good](https://github.com/djkoloski/rust_serialization_benchmark) library.

1. Include [rmp-serde](https://crates.io/crates/rmp-serde) in your project.
2. include the following struct and trait implementations in your project:
```rust
pub struct RmpSerde;

impl<T: serde::Serialize> native_model::Encode<T> for RmpSerde {
    type Error = rmp_serde::encode::Error;
    fn encode(obj: &T) -> Result<Vec<u8>, Self::Error> {
        rmp_serde::encode::to_vec(obj)
    }
}

impl<T: for<'a> serde::Deserialize<'a>> native_model::Decode<T> for RmpSerde {
    type Error = rmp_serde::decode::Error;
    fn decode(data: Vec<u8>) -> Result<T, Self::Error> {
        rmp_serde::decode::from_slice(&data)
    }
}
```
3. Apply the following derive macro to the struct you'd like to use with `native_db`:
```rust
#[native_model(id = 1, version = 1, with = RmpSerde)]
struct My Struct {
    pub string: String,
}
```

@vincent-herlemont  Yes I will try to make a PR sometime this weekend

**Comment by LunaeSomnia** [comment]
Hey @leontoeides, I just implemented your instructions & ended up working. The serialization made by `rmp-serde` is surprisingly quick and I'll be using it from now on in my project.

Thank you very much for your time.

**Comment by vincent-herlemont** [comment]
With the merging of this PR https://github.com/vincent-herlemont/native_db/pull/141, the database will no longer cause a panic related to deserialization. Each query returns a result. This will be available in version 0.7.0 of `native_db`.
